src/storage/buffer_manager.cpp

```diff
@@ -173,6 +173,37 @@ shared_ptr<BlockHandle> BufferManager::RegisterBlock(block_id_t block_id) {
 	return result;
 }
 
+shared_ptr<BlockHandle> BufferManager::ConvertToPersistent(BlockManager &block_manager, block_id_t block_id,
+                                                           shared_ptr<BlockHandle> old_block) {
+	// pin the old block to ensure we have it loaded in memory
+	auto old_handle = Pin(old_block);
+	D_ASSERT(old_block->state == BlockState::BLOCK_LOADED);
+	D_ASSERT(old_block->buffer);
+
+	// register a block with the new block id
+	auto new_block = RegisterBlock(block_id);
+	D_ASSERT(new_block->state == BlockState::BLOCK_UNLOADED);
+	D_ASSERT(new_block->readers == 0);
+
+	// move the data from the old block into data for the new block
+	new_block->state = BlockState::BLOCK_LOADED;
+	new_block->buffer = make_unique<Block>(*old_block->buffer, block_id);
+
+	// clear the old buffer and unload it
+	old_handle.reset();
```

Comment:
The pin holds a raw pointer to the contents of the block, so if we release the block contents first the pin will hold a stale pointer to the block contents.

It doesn't really matter either way since we are not using the pointer after this point.